### DISCUSSION TITLE
How to find certain source codes for a function

### DISCUSSION THREAD

**Discussion by LanceKnight** [rdkit/rdkit#6839] [FAQ]
Hello, I was wondering how rdkit calculate the rotatable bonds (what criteria it uses to classify a bond into a rotatable bond). So I was trying to read the Chem.Linpiski.NumRotatableBonds. However, I encountered several roadblocks:

1. I tried to search it in the [C++ reference](https://www.rdkit.org/docs/cppapi/index.html) but I couldn't find the class Chem in the left panel where it says "Classes-Class List"
2. I then found Chem.Linpiski.NumRotatableBonds in the github repo. Then inside the rdkit/Chem folder in the repo, I found it uses Chem.rdMolDescriptors. However, I was not able to find a file or folder called rdMolDescriptors in the Chem folder in the github

Can someone teach how to, in general, look for source codes for certain things in rdkit? I'm more used to pytorch type of documentation where [source codes] is listed right next to the function in the documentation.

**Comment by greglandrum** [comment]
Hi @LanceKnight,

As you've discovered, it can be a bit complicated to find things like this due to the fact that most of the RDKit is implemented in C++ which is then wrapped to be used from Python. The names of things in the C++ code are usually not exactly the same as the names in Python.

Probably the easiest way to answer a question like "how is the number of rotatable bonds defined" is to search in the rdkit repo in github ![image](https://github.com/rdkit/rdkit/assets/540511/f9626780-f9cf-4898-a932-7ee2a8d50936)
for `NumRotatableBonds`:
![image](https://github.com/rdkit/rdkit/assets/540511/b133d0eb-f246-41aa-b454-cc40f52f1f77)
Then look through the C++ results until you find something that looks like it might be the answer;
![image](https://github.com/rdkit/rdkit/assets/540511/d46d4af4-18d5-48c4-a784-322433b02d10)
Click on that to go to the right file, and then you can find the actual function that implements the descriptor:
https://github.com/rdkit/rdkit/blob/4a69bc3493dd3e9bb9f7a519e306fbcb545f1452/Code/GraphMol/Descriptors/Lipinski.cpp#L108

The other piece of information you need in this case is that the default value of the strict argument is `Strict`.

**Reply by YunchaoLiu** [reply]
I see. Thank you so much for your detailed answer @greglandrum :)

**Reply by YunchaoLiu** [reply]
btw, if there is no rdMolDescriptor folder/file under Chem folder, how does this import realized? I guess it's also related to the C++ implementation? Sorry I really have little experience with C++ with python wrapping so I am not sure how that works
![image](https://github.com/rdkit/rdkit/assets/41695702/de15bdc3-e7ae-40ca-ae33-4f8f9fd4452d)

**Reply by bp-kelley** [reply]
Generally everything that starts with an “rd” is a shared library that python can import.
